### PR TITLE
Add stable diffusion fine-tuning

### PR DESCRIPTION
**Pull request by itrushkin**
### Summary
This PR introduces a new "Diffusion" task for image generation using LDMs.

### Checklist

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [x] I have ran e2e tests and there is no issues.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [x] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [x] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by yunchu** [comment]
@itrushkin please update base branch to apply #3882  it will resolve storage related issue on the unittest.

File: pyproject.toml [hunk 1]
```diff
@@ -106,6 +106,8 @@ base = [
 
 transformers = [
     "transformers==4.44.1",
+    "diffusers==0.30.0",
+    "torchmetrics[image]>=0.7.0"
```
Comment (wonjuleee) [inline pyproject.toml:108]: Is there no conflict with lightening torchmetric?

File: src/otx/algo/diffusion/clip_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
+"""This module provides the ClipTokenizer class for tokenizing text using the ClipTokenizer algorithm."""
```
Comment (wonjuleee) [inline src/otx/algo/diffusion/clip_tokenizer.py]: Please add Intel copyright :)

File: src/otx/algo/diffusion/clip_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+from __future__ import annotations
+
+import gzip
+import re
+from functools import lru_cache
+
+from .utils.download import download
+
+
+@lru_cache
```
Comment (wonjuleee) [inline src/otx/algo/diffusion/clip_tokenizer.py]: Should we have `lru_cache` for this? This looks like called once when creating Tokenizer class.

File: pyproject.toml (hunk 1 above)
Comment (itrushkin) [reply pyproject.toml:108]: This version specification is the same as in `pytorch_lightning`

File: src/otx/algo/diffusion/huggingface_model.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        outputs: torch.Tensor,
+        inputs: DiffusionBatchDataEntity,
+    ) -> DiffusionBatchPredEntity | OTXBatchLossEntity:
+        if torch.isnan(outputs).any():
+            error_msg = "NaN detected in the output."
+            raise ValueError(error_msg)
+        if not self.training:
+            msg = "This should never raise since `validation_step` is overridden."
+            raise NotImplementedError(msg)
+        return OTXBatchLossEntity(mse=F.mse_loss(outputs, self.target, reduction="mean"))
```
Comment (sungchul1) [inline src/otx/algo/diffusion/huggingface_model.py]: `_customize_outputs` is for converting or wrapping outputs to `OTXBatchPredEntity` or `OTXBatchLossEntity`, not calculating loss or post processing.
Could you add this loss calculation in the model if possible?

File: src/otx/algo/diffusion/huggingface_model.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            do_denormalize=[True] * len(images),
+        )
+        self.metric.update(images, real=False)
+
+        # Save images
+        for logger in self.loggers:
+            if isinstance(logger, TensorBoardLogger):
+                for image, caption in zip(images, batch.captions):
+                    logger.experiment.add_image(f"val/images/{caption}", image.detach().cpu())
+        return images
```
Comment (sungchul1) [inline src/otx/algo/diffusion/huggingface_model.py]: I don't know much, but doesn't diffusion have any metric calculation?
As you already know, `validation_step` and `test_step` are for metric calculation.

File: src/otx/algo/diffusion/utils/data_sample.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+"""Implementation of diffusion data sample."""
+
+from __future__ import annotations
+
+import numpy as np
+from otx.algo.utils.mmengine_utils import InstanceData
+
+
+class DiffusionDataSample(InstanceData):
```
Comment (sungchul1) [inline src/otx/algo/diffusion/utils/data_sample.py]: `InstanceData` will be deprecated, I'd suggest not using `InstanceData` for the base class.
And if there is no reason to use this data container, could you use data entity instead?

File: src/otx/core/exporter/diffusion.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+        return {"model": fn(model, output_dir, base_model_name, precision)}
+
+    def to_openvino(
+        self,
+        model: OTXModel | torch.nn.Module,
+        output_dir: Path,
+        base_model_name: str = "exported_model",
+        precision: OTXPrecisionType = OTXPrecisionType.FP32,
+        model_type: str = "sam",
```
Comment (sungchul1) [inline src/otx/core/exporter/diffusion.py]: This code looks like it was taken from a visual prompt and hasn't been updated yet?

File: src/otx/algo/diffusion/huggingface_model.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+    def test_step(self, batch: DiffusionBatchDataEntity, batch_idx: int) -> None:
+        """Perform a single test step on a batch of data from the test set.
+
+        :param batch: A batch of data (a tuple) containing the input tensor of images and target
+            labels.
+        :param batch_idx: The index of the current batch.
+        """
+        return self.validation_step(batch, batch_idx)
```
Comment (sungchul1) [inline src/otx/algo/diffusion/huggingface_model.py]: It seems similar with `OTXDiffusionModel`.

File: src/otx/recipe/diffusion/sd_huggingface.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+engine:
+  task: DIFFUSION
+  device: auto
+
+callback_monitor: val/fid
+
+data: ../_base_/data/diffusion.yaml
+max_steps: 60000
+min_steps: 60000
```
Comment (sungchul1) [inline src/otx/recipe/diffusion/sd_huggingface.yaml:27]: Are these parameters related to training?

File: tests/assets/naruto_blip_captions/images/val/ea15e617.jpg
Comment (sungchul1) [inline tests/assets/naruto_blip_captions/images/val/ea15e617.jpg:1]: Don't these images for test have any license or copyright issue?

**Review by sungchul1 — COMMENTED** [review]
Thanks for your hard work.
I left some comments, please take a look.
And I'd like to request two things:
1. please check all docstrings in the files you worked. There are such things that are not updated, incorrect formats, and not aligned with arguments.
2. could you update performance tables to check reproducibility with official performance and torch - ov performance consistency.

File: tests/assets/naruto_blip_captions/images/val/ea15e617.jpg
Comment (itrushkin) [reply tests/assets/naruto_blip_captions/images/val/ea15e617.jpg:1]: The images are taken from the [HuggingFace dataset](https://huggingface.co/datasets/lambdalabs/naruto-blip-captions). There is no license specified, but if there is any issue, all images can be replaced with the black ones.

File: src/otx/recipe/diffusion/sd_huggingface.yaml (hunk 9 above)
Comment (itrushkin) [reply src/otx/recipe/diffusion/sd_huggingface.yaml:27]: Yes, these parameters are passed as `Trainer` keyword arguments and describe the amount of steps like `max_epochs` and `min_epochs`. [docs link](https://lightning.ai/docs/pytorch/stable/common/trainer.html#max-steps)

File: src/otx/algo/diffusion/huggingface_model.py (hunk 5 above)
Comment (itrushkin) [reply src/otx/algo/diffusion/huggingface_model.py]: For Diffusion task, Frechet Inception Distance (FID) metric is used. It is updated via `metric.update()` call in `validation_step` method.

File: src/otx/algo/diffusion/huggingface_model.py (hunk 4 above)
Comment (itrushkin) [reply src/otx/algo/diffusion/huggingface_model.py]: Done.

File: src/otx/algo/diffusion/utils/data_sample.py (hunk 6 above)
Comment (itrushkin) [reply src/otx/algo/diffusion/utils/data_sample.py]: Done.

File: src/otx/core/exporter/diffusion.py (hunk 7 above)
Comment (itrushkin) [reply src/otx/core/exporter/diffusion.py]: Updated.

File: src/otx/algo/diffusion/huggingface_model.py (hunk 8 above)
Comment (itrushkin) [reply src/otx/algo/diffusion/huggingface_model.py]: Removed the `OTXDiffusionModel` duplicate.

File: src/otx/core/exporter/diffusion.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        """
+        if export_format == OTXExportFormatType.OPENVINO:
+            if to_exportable_code:
+                msg = "Exportable code option is not supported and will be ignored."
+                log.warning(msg)
+            fn = self.to_openvino
+        else:
+            fn = self.to_onnx  # type: ignore[assignment]
+
+        return {"model": fn(model, output_dir, base_model_name, precision)}
```
Comment (sungchul1) [inline src/otx/core/exporter/diffusion.py]: Is it necessary? `export` seems from `OTXModelExporter` in base.py and it just returns outputs not using dict.

File: src/otx/algo/diffusion/huggingface_model.py [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+            "encoder_hidden_states": encoder_hidden_states,
+            "target_noise": target,
+        }
+
+    def _customize_outputs(
+        self,
+        outputs: torch.Tensor,
+        inputs: DiffusionBatchDataEntity,
+    ) -> DiffusionBatchPredEntity | OTXBatchLossEntity:
+        return OTXBatchLossEntity(loss=outputs)
```
Comment (sungchul1) [inline src/otx/algo/diffusion/huggingface_model.py]: It seems better to move into `core/model/diffusion` because it can be commonly used with other diffucision models to be added. What do you think?

File: src/otx/algo/diffusion/huggingface_model.py [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+                    logger.experiment.add_image(f"val/images/{caption}", image.detach().cpu())
+
+    def test_step(self, batch: DiffusionBatchDataEntity, batch_idx: int) -> None:
+        """Perform a single test step on a batch of data from the test set.
+
+        :param batch: A batch of data (a tuple) containing the input tensor of images and target
+            labels.
+        :param batch_idx: The index of the current batch.
+        """
+        self.validation_step(batch, batch_idx)
```
Comment (sungchul1) [inline src/otx/algo/diffusion/huggingface_model.py]: Same with above, because there is no `self.pipe` here.

**Review by sungchul1 — COMMENTED** [review]
Thanks for updates.
I left some comments and I'd request to leave commit id in each comment to check update easier :D

**Comment by kprokofi** [comment]
Hi @itrushkin , if this PR intends to be included in OTX2.3 release, it should be merged today. Otherwise, we need to postpone it for 2.4

File: src/otx/core/exporter/diffusion.py (hunk 10 above)
Comment (itrushkin) [reply src/otx/core/exporter/diffusion.py]: This class inherits from `OTXNativeModelExporter` which uses `onnx_export_configuration` as an export parameter for ONNX. `export` method in this parent class returns `dict[str, Path]`, so the child class method should have the same return type.

File: src/otx/algo/diffusion/huggingface_model.py (hunk 12 above)
Comment (itrushkin) [reply src/otx/algo/diffusion/huggingface_model.py]: Fixed in [4a82d0d](https://github.com/openvinotoolkit/training_extensions/pull/3875/commits/4a82d0d2aa2f5b16803f10ac9f2d1725dd542682)

File: src/otx/algo/diffusion/huggingface_model.py (hunk 11 above)
Comment (itrushkin) [reply src/otx/algo/diffusion/huggingface_model.py]: Fixed in [4a82d0d](https://github.com/openvinotoolkit/training_extensions/pull/3875/commits/4a82d0d2aa2f5b16803f10ac9f2d1725dd542682)

File: src/otx/core/exporter/diffusion.py (hunk 10 above)
Comment (sungchul1) [reply src/otx/core/exporter/diffusion.py]: No, OTXNativeModelExporter inherits export method from OTXModelExporter and its return type is Path, not dict[str, Path].
https://github.com/openvinotoolkit/training_extensions/blob/0387d8bcd629418a3942e2469c2ba2139560b696/src/otx/core/exporter/base.py#L78-L113

Comment (itrushkin) [reply src/otx/core/exporter/diffusion.py]: Fixed in [457ccd0](https://github.com/openvinotoolkit/training_extensions/pull/3875/commits/457ccd07ef8f422ed039aecea7f25aa0b30befa1)

**Review by wonjuleee — DISMISSED** [review]

**Review by kprokofi — CHANGES_REQUESTED** [review]
postpone to merge PR before 2.2/2.3 is finalized

**Review by kprokofi — DISMISSED** [review]

**Review by wonjuleee — APPROVED** [review]

**Review by sungchul1 — APPROVED** [review]
@itrushkin LGTM, could you add unit tests for each files/components you added/updated in the next PR?
